File: packages/grua_core/test/zone_pricing_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';
import 'package:grua_core/grua_core.dart';

/// The insurer zone tariff, run against the examples the server also runs.
///
/// `functions/test/zonePricing.test.ts` reads the same file, so a price that
/// changes on one side only fails a test on the other.
void main() {
  final fixture = jsonDecode(
    File('test/fixtures/zone_pricing_cases.json').readAsStringSync(),
  ) as Map<String, dynamic>;

  List<PricingRule> rulesOf(Object? json) => [
        for (final row in json! as List<dynamic>)
          PricingRule.fromJson(row as Map<String, dynamic>),
      ];

  VehicleType vehicle(String wire) => VehicleType.fromWire(wire);

  group('the default table', () {
    test('is the table in the shared examples', () {
      expect(ZonePricing.defaultRules, rulesOf(fixture['defaultRules']));
    });

    test('is complete for every priced class', () {
      for (final vehicleClass in VehicleClass.priced) {
        expect(
          ZonePricing.tableProblem(ZonePricing.defaultRulesFor(vehicleClass)),
          isNull,
          reason: vehicleClass.wire,
        );
      }
    });

    test('has a row id per zone, matching the server', () {
      final ids = ZonePricing.defaultRules.map((r) => r.id).toSet();
      expect(ids, hasLength(ZonePricing.defaultRules.length));
      expect(ZonePricing.defaultRules.first.id, 'default__light__0');
    });

    test('labels zones the way the price list does', () {
      final light = ZonePricing.defaultRulesFor(VehicleClass.light);
      expect(light.first.zoneLabel, '0–10 km');
      expect(light.last.zoneLabel, '+50 km');
    });
  });

  group('VehicleClass.of', () {
    final classes = fixture['vehicleClasses'] as Map<String, dynamic>;
    for (final entry in classes.entries) {
      test('prices ${entry.key} as ${entry.value ?? 'nothing'}', () {
        final expected = entry.value == null
            ? VehicleClass.unknown
            : VehicleClass.fromWire(entry.value as String);
        expect(VehicleClass.of(vehicle(entry.key)), expected);
      });
    }

    test('every vehicle type the app offers has a column', () {
      for (final type in VehicleType.values) {
        if (type == VehicleType.unknown) continue;
        expect(VehicleClass.priced, contains(VehicleClass.of(type)), reason: type.wire);
      }
    });
  });

  group('ZonePricing.quote on the default table', () {
    for (final raw in fixture['defaultCases'] as List<dynamic>) {
      final c = raw as Map<String, dynamic>;
      test(c['name'] as String, () {
        final vehicleClass = VehicleClass.of(vehicle(c['vehicleType'] as String));
        final quote = ZonePricing.quote(
          rules: ZonePricing.defaultRulesFor(vehicleClass),
          distanceKm: (c['distanceKm'] as num).toDouble(),
          tariff: ZoneTariffSource.standard,
        );

        expect(quote.vehicleClass, vehicleClass);
        expect(quote.zoneMinKm, c['zoneMinKm']);
        expect(quote.distanceKm, (c['distanceOut'] as num).toDouble());
        expect(quote.extraKm, (c['extraKm'] as num).toDouble());
        expect(quote.extraCents, c['extraCents']);
        expect(quote.subtotalCents, c['subtotalCents']);
        expect(quote.tariff.wire, 'default');
      });
    }

    test('refuses a distance that is not a distance', () {
      final rules = ZonePricing.defaultRulesFor(VehicleClass.light);
      for (final d in [-1.0, double.nan, double.infinity]) {
        expect(
          () => ZonePricing.quote(rules: rules, distanceKm: d, tariff: ZoneTariffSource.standard),
          throwsRangeError,
          reason: '$d',
        );
      }
    });
  });

  group('ZonePricing.quote on other tables', () {
    for (final rawTable in fixture['customTables'] as List<dynamic>) {
      final table = rawTable as Map<String, dynamic>;
      for (final raw in table['cases'] as List<dynamic>) {
        final c = raw as Map<String, dynamic>;
        test('${table['name']}: ${c['distanceKm']} km', () {
          final quote = ZonePricing.quote(
            rules: rulesOf(table['rules']),
            distanceKm: (c['distanceKm'] as num).toDouble(),
            tariff: ZoneTariffSource.insurer,
          );
          expect(quote.zoneMinKm, c['zoneMinKm']);
          expect(quote.extraCents, c['extraCents']);
          expect(quote.subtotalCents, c['subtotalCents']);
        });
      }
    }
  });

  group('ZonePricing.tableProblem', () {
    for (final raw in fixture['invalidTables'] as List<dynamic>) {
      final table = raw as Map<String, dynamic>;
      test('refuses a table: ${table['name']}', () {
        final rules = rulesOf(table['rules']);
        expect(ZonePricing.tableProblem(rules), isA<String>());
        expect(
          () => ZonePricing.quote(
            rules: rules,
            distanceKm: 5,
            tariff: ZoneTariffSource.standard,
          ),
          throwsA(isA<ZoneTableException>()),
        );
      });
    }
  });

  group('ZonePricing.withItbis', () {
    for (final raw in fixture['itbis'] as List<dynamic>) {
      final c = raw as Map<String, dynamic>;
      test('${c['subtotalCents']} + 18%', () {
        final totals = ZonePricing.withItbis(c['subtotalCents'] as int);
        expect(totals.itbisCents, c['itbisCents']);
        expect(totals.totalCents, c['totalCents']);
      });
    }
  });

  group('ZonePricing.driverPayoutCents', () {
    test('defaults to 70%', () {
      expect(ZonePricing.defaultDriverPayoutBps, 7000);
    });

    for (final raw in fixture['payouts'] as List<dynamic>) {
      final c = raw as Map<String, dynamic>;
      test('${c['driverPayoutBps']} bps of ${c['subtotalCents']}', () {
        expect(
          ZonePricing.driverPayoutCents(
            c['subtotalCents'] as int,
            c['driverPayoutBps'] as int,
          ),
          c['driverPayoutCents'],
        );
      });
    }
  });

  group('PricingRule', () {
    test('reads what it writes', () {
      for (final rule in ZonePricing.defaultRules) {
        expect(PricingRule.fromJson(rule.toJson()), rule);
      }
    });

    test('an unknown class reads as unknown and cannot be priced', () {
      final rule = PricingRule.fromJson(const {
        'vehicleClass': 'moto',
        'zoneMinKm': 0,
        'zoneMaxKm': null,
        'baseCents': 1,
        'extraKmCents': 0,
        'insurerId': null,
      });
      expect(rule.vehicleClass, VehicleClass.unknown);
      expect(VehicleClass.priced, isNot(contains(rule.vehicleClass)));
    });
  });

  test('the collection name matches the backend', () {
    expect(Paths.pricingRulesCollection, 'pricingRules');
  });
}

File: functions/test/zonePricing.test.ts
import { readFileSync } from 'node:fs';

import { describe, expect, it } from 'vitest';

import {
  DEFAULT_PRICING_RULES,
  type PricingRule,
  ZoneTableError,
  defaultRulesFor,
  pricingRuleId,
  pricingRuleSchema,
  quoteZonePrice,
  vehicleClassOf,
  withItbis,
  zoneTableProblem,
} from '../src/lib/zonePricing.js';

/**
 * The insurer zone tariff, run against the examples the Dart port also runs.
 * A number that changes here without changing there fails one of the two.
 */

interface Fixture {
  defaultRules: PricingRule[];
  vehicleClasses: Record<string, string | null>;
  defaultCases: Array<{
    name: string;
    vehicleType: string;
    distanceKm: number;
    zoneMinKm: number;
    distanceOut: number;
    extraKm: number;
    extraCents: number;
    subtotalCents: number;
  }>;
  customTables: Array<{
    name: string;
    rules: PricingRule[];
    cases: Array<{ distanceKm: number; zoneMinKm: number; extraCents: number; subtotalCents: number }>;
  }>;
  invalidTables: Array<{ name: string; rules: PricingRule[] }>;
  itbis: Array<{ subtotalCents: number; itbisCents: number; totalCents: number }>;
}

const fixture = JSON.parse(
  readFileSync(
    new URL('../../packages/grua_core/test/fixtures/zone_pricing_cases.json', import.meta.url),
    'utf8',
  ),
) as Fixture;

describe('the default table', () => {
  it('is the table in the shared examples', () => {
    expect([...DEFAULT_PRICING_RULES]).toEqual(fixture.defaultRules);
  });

  it('is a complete table for every class', () => {
    for (const vehicleClass of ['light', 'suv', 'heavy'] as const) {
      expect(zoneTableProblem(defaultRulesFor(vehicleClass))).toBeNull();
    }
  });

  it('holds only rows a pricingRules document may hold', () => {
    for (const rule of DEFAULT_PRICING_RULES) {
      expect(pricingRuleSchema.safeParse(rule).success).toBe(true);
    }
  });

  it('matches the printed price list', () => {
    const at = (vehicleClass: 'light' | 'suv' | 'heavy', zoneMinKm: number) =>
      DEFAULT_PRICING_RULES.find(
        (r) => r.vehicleClass === vehicleClass && r.zoneMinKm === zoneMinKm,
      )!;
    // "Tabla de precios base — Titan Grúas RD (sin ITBIS)", in pesos.
    expect(at('light', 0).baseCents / 100).toBe(2_500);
    expect(at('suv', 10).baseCents / 100).toBe(4_500);
    expect(at('heavy', 25).baseCents / 100).toBe(11_000);
    expect(at('light', 50).extraKmCents / 100).toBe(120);
    expect(at('suv', 50).extraKmCents / 100).toBe(150);
    expect(at('heavy', 50).extraKmCents / 100).toBe(250);
  });
});

describe('vehicleClassOf', () => {
  for (const [vehicleType, vehicleClass] of Object.entries(fixture.vehicleClasses)) {
    it(`prices ${vehicleType} as ${vehicleClass ?? 'nothing'}`, () => {
      expect(vehicleClassOf(vehicleType)).toBe(vehicleClass);
    });
  }

  it('has no column for a value it does not know', () => {
    expect(vehicleClassOf('bicicleta')).toBeNull();
    expect(vehicleClassOf('')).toBeNull();
  });
});

describe('quoteZonePrice on the default table', () => {
  for (const c of fixture.defaultCases) {
    it(c.name, () => {
      const vehicleClass = vehicleClassOf(c.vehicleType)!;
      const quote = quoteZonePrice({
        rules: defaultRulesFor(vehicleClass),
        distanceKm: c.distanceKm,
        tariff: 'default',
      });

      expect(quote.vehicleClass).toBe(vehicleClass);
      expect(quote.zoneMinKm).toBe(c.zoneMinKm);
      expect(quote.distanceKm).toBe(c.distanceOut);
      expect(quote.extraKm).toBe(c.extraKm);
      expect(quote.extraCents).toBe(c.extraCents);
      expect(quote.subtotalCents).toBe(c.subtotalCents);
      expect(quote.baseCents + quote.extraCents).toBe(quote.subtotalCents);
      expect(quote.tariff).toBe('default');
      expect(quote.currency).toBe('DOP');
    });
  }

  it('never charges less for a longer trip', () => {
    for (const vehicleClass of ['light', 'suv', 'heavy'] as const) {
      let previous = 0;
      for (let tenths = 0; tenths <= 2000; tenths++) {
        const { subtotalCents } = quoteZonePrice({
          rules: defaultRulesFor(vehicleClass),
          distanceKm: tenths / 10,
          tariff: 'default',
        });
        expect(subtotalCents).toBeGreaterThanOrEqual(previous);
        previous = subtotalCents;
      }
    }
  });

  it('charges heavy more than SUV, and SUV more than light, at every distance', () => {
    for (let tenths = 0; tenths <= 2000; tenths += 7) {
      const price = (vehicleClass: 'light' | 'suv' | 'heavy') =>
        quoteZonePrice({
          rules: defaultRulesFor(vehicleClass),
          distanceKm: tenths / 10,
          tariff: 'default',
        }).subtotalCents;
      expect(price('suv')).toBeGreaterThan(price('light'));
      expect(price('heavy')).toBeGreaterThan(price('suv'));
    }
  });

  it('does not care what order the rows come in', () => {
    const shuffled = [...defaultRulesFor('light')].reverse();
    const quote = quoteZonePrice({ rules: shuffled, distanceKm: 62, tariff: 'default' });
    expect(quote.subtotalCents).toBe(694_000);
  });

  it('refuses a distance that is not a distance', () => {
    const rules = defaultRulesFor('light');
    for (const distanceKm of [-1, Number.NaN, Number.POSITIVE_INFINITY]) {
      expect(() => quoteZonePrice({ rules, distanceKm, tariff: 'default' })).toThrow(RangeError);
    }
  });
});

describe('quoteZonePrice on other tables', () => {
  for (const table of fixture.customTables) {
    for (const c of table.cases) {
      it(`${table.name}: ${c.distanceKm} km`, () => {
        const quote = quoteZonePrice({
          rules: table.rules,
          distanceKm: c.distanceKm,
          tariff: 'insurer',
        });
        expect(quote.zoneMinKm).toBe(c.zoneMinKm);
        expect(quote.extraCents).toBe(c.extraCents);
        expect(quote.subtotalCents).toBe(c.subtotalCents);
        expect(quote.tariff).toBe('insurer');
      });
    }
  }
});

describe('zoneTableProblem', () => {
  for (const table of fixture.invalidTables) {
    it(`refuses a table: ${table.name}`, () => {
      expect(zoneTableProblem(table.rules)).toEqual(expect.any(String));
      expect(() =>
        quoteZonePrice({ rules: table.rules, distanceKm: 5, tariff: 'default' }),
      ).toThrow(ZoneTableError);
    });
  }

  it('accepts a single open zone', () => {
    expect(zoneTableProblem(fixture.customTables[1]!.rules)).toBeNull();
  });
});

describe('pricingRuleSchema', () => {
  const valid = fixture.defaultRules[0]!;

  it('refuses fractional kilometres and cents', () => {
    expect(pricingRuleSchema.safeParse({ ...valid, zoneMinKm: 0.5 }).success).toBe(false);
    expect(pricingRuleSchema.safeParse({ ...valid, baseCents: 10.5 }).success).toBe(false);
  });

  it('refuses negative prices', () => {
    expect(pricingRuleSchema.safeParse({ ...valid, baseCents: -1 }).success).toBe(false);
    expect(pricingRuleSchema.safeParse({ ...valid, extraKmCents: -1 }).success).toBe(false);
  });

  it('refuses a class that is not in the table', () => {
    expect(pricingRuleSchema.safeParse({ ...valid, vehicleClass: 'moto' }).success).toBe(false);
  });

  it('requires insurerId to be present, even when it is null', () => {
    const { insurerId: _, ...rest } = valid;
    expect(pricingRuleSchema.safeParse(rest).success).toBe(false);
  });
});

describe('pricingRuleId', () => {
  it('names the table, the class and the zone', () => {
    expect(pricingRuleId(fixture.defaultRules[0]!)).toBe('default__light__0');
    expect(pricingRuleId(fixture.customTables[0]!.rules[1]!)).toBe('ins-demo__light__10');
  });

  it('is different for every default row', () => {
    const ids = new Set(DEFAULT_PRICING_RULES.map(pricingRuleId));
    expect(ids.size).toBe(DEFAULT_PRICING_RULES.length);
  });
});

describe('withItbis', () => {
  for (const c of fixture.itbis) {
    it(`RD$ cents ${c.subtotalCents} + 18%`, () => {
      expect(withItbis(c.subtotalCents)).toEqual(c);
    });
  }
});
